Parse unnamed fields and anonymous structs or unions

frank-king opened the pull request:
This PR implements #49804, and parses the anonymous structs or unions as a special type of field, instead of a general kind of `Ty`.

r? @petrochenkov

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6138086544/job/16654607709) [(plain)](https://github.com/rust-lang/rust/commit/36df386d5efe5c1b2b3d11f8a8e61d065fb55ca9/checks/16654607709/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=frank-king
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_18387680-f0f8-4012-808d-299941caa292
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=feature/unnamed-fields-lite
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_18387680-f0f8-4012-808d-299941caa292
GITHUB_REF=refs/pull/115732/merge
GITHUB_REF_NAME=115732/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=2bf71d4bcf9d0e21ccd570f7635a91ef7f2a4577
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_18387680-f0f8-4012-808d-299941caa292
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_18387680-f0f8-4012-808d-299941caa292
GITHUB_TRIGGERING_ACTOR=frank-king
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115732/merge
GITHUB_WORKFLOW_SHA=2bf71d4bcf9d0e21ccd570f7635a91ef7f2a4577
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
   Compiling declare_clippy_lint v0.1.74 (/checkout/src/tools/clippy/declare_clippy_lint)
error[E0308]: arguments to this function are incorrect
   --> src/tools/clippy/clippy_utils/src/ast_utils.rs:559:12
    |
559 |         && eq_ty(&l.ty, &r.ty)
    |            ^^^^^ -----  ----- expected `&Ty`, found `&FieldTy`
    |                  |
    |                  expected `&Ty`, found `&FieldTy`
    = note: expected reference `&rustc_ast::Ty`
    = note: expected reference `&rustc_ast::Ty`
               found reference `&FieldTy`
    = note: expected reference `&rustc_ast::Ty`
               found reference `&FieldTy`
   --> src/tools/clippy/clippy_utils/src/ast_utils.rs:659:8
    |
    |
659 | pub fn eq_ty(l: &Ty, r: &Ty) -> bool {

   Compiling futures-macro v0.3.28
    Checking tracing-error v0.2.0
For more information about this error, try `rustc --explain E0308`.
```

</details>

nnethercote commented on compiler/rustc_ast/src/ast.rs:
```diff
@@ -2765,6 +2869,44 @@ impl VariantData {
 
+/// Iterator of all fields of a `VariantData`.
+///
+/// It iteartes on the field tree in preorder, where the unnamed fields with anonymous structs or unions
```
```suggestion
/// It iterates on the field tree in preorder, where the unnamed fields with anonymous structs or unions
```

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6148775513/job/16683229691) [(plain)](https://github.com/rust-lang/rust/commit/13209614b9868f7da9279f2c574860b4196a5bcc/checks/16683229691/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=frank-king
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_c80e1efa-bf24-4f48-9a42-89beff3730d5
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=feature/unnamed-fields-lite
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_c80e1efa-bf24-4f48-9a42-89beff3730d5
GITHUB_REF=refs/pull/115732/merge
GITHUB_REF_NAME=115732/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=b1a7fe53bbb36f6f2554c8ee60a5ff44f4c64c99
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_c80e1efa-bf24-4f48-9a42-89beff3730d5
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_c80e1efa-bf24-4f48-9a42-89beff3730d5
GITHUB_TRIGGERING_ACTOR=frank-king
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115732/merge
GITHUB_WORKFLOW_SHA=b1a7fe53bbb36f6f2554c8ee60a5ff44f4c64c99
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
   Compiling derive-new v0.5.9
error[E0609]: no field `kind` on type `FieldTy`
   --> src/tools/clippy/clippy_lints/src/manual_non_exhaustive.rs:119:29
    |
119 |                 && field.ty.kind.is_unit()

For more information about this error, try `rustc --explain E0609`.
error: could not compile `clippy_lints` (lib) due to previous error
Build completed unsuccessfully in 0:02:56
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6158692817/job/16711977871) [(plain)](https://github.com/rust-lang/rust/commit/a05f6d39cc44263d2977a818719071e4124e15be/checks/16711977871/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=frank-king
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_73879124-3850-4c9f-9efa-752603d8b94d
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=feature/unnamed-fields-lite
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_73879124-3850-4c9f-9efa-752603d8b94d
GITHUB_REF=refs/pull/115732/merge
GITHUB_REF_NAME=115732/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=afc8e3fd095d37b053581ae6cdd101ba59181ffc
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_73879124-3850-4c9f-9efa-752603d8b94d
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_73879124-3850-4c9f-9efa-752603d8b94d
GITHUB_TRIGGERING_ACTOR=frank-king
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115732/merge
GITHUB_WORKFLOW_SHA=afc8e3fd095d37b053581ae6cdd101ba59181ffc
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
    Checking toml_edit v0.19.11
    Checking serde_json v1.0.99
    Checking cargo_metadata v0.15.4
    Checking toml v0.7.5
error[E0599]: no method named `rewrite` found for enum `FieldTy` in the current scope
     |
     |
1892 |         .and_then(|ty_shape| field.ty.rewrite(context, ty_shape));
     |                                       ^^^^^^^ method not found in `FieldTy`
     |
note: the method `rewrite` exists on the type `P<rustc_ast::Ty>`
     |
     |
18   |     fn rewrite(&self, context: &RewriteContext<'_>, shape: Shape) -> Option<String>;
     = help: items from traits can only be used if the trait is implemented and in scope
     = help: items from traits can only be used if the trait is implemented and in scope
note: `Rewrite` defines an item `rewrite`, perhaps you need to implement it
     |
     |
16   | pub(crate) trait Rewrite {

error[E0282]: type annotations needed for `std::option::Option<U>`
    --> src/tools/rustfmt/src/items.rs:1890:9
     |
     |
1890 |     let orig_ty = shape
     |         ^^^^^^^
...
1894 |         if !ty.contains('\n') && !contains_comment(context.snippet(missing_span)) {
     |                -------- type must be known at this point
     |
help: consider giving `orig_ty` an explicit type, where the type for type parameter `T` is specified
     |
1890 |     let orig_ty: std::option::Option<U> = shape


error[E0614]: type `FieldTy` cannot be dereferenced
     |
     |
1901 |     let field_str = rewrite_assign_rhs(context, prefix, &*field.ty, &RhsAssignKind::Ty, shape)?;


error[E0615]: attempted to take value of method `span` on type `FieldTy`
    |
    |
141 |         span_with_attrs_lo_hi!(self, self.span.lo(), self.ty.span.hi())
    |
help: use parentheses to call the method
    |
    |
141 |         span_with_attrs_lo_hi!(self, self.span.lo(), self.ty.span().hi())


error[E0599]: no variant or associated item named `AnonStruct` found for enum `TyKind` in the current scope
    |
    |
822 |             ast::TyKind::AnonStruct(_) => Some(context.snippet(self.span).to_owned()),
    |                          ^^^^^^^^^^ variant or associated item not found in `TyKind`

error[E0599]: no variant or associated item named `AnonUnion` found for enum `TyKind` in the current scope
    |
    |
823 |             ast::TyKind::AnonUnion(_) => Some(context.snippet(self.span).to_owned()),
    |                          ^^^^^^^^^ variant or associated item not found in `TyKind`
Some errors have detailed explanations: E0282, E0599, E0614, E0615.
For more information about an error, try `rustc --explain E0282`.
error: could not compile `rustfmt-nightly` (lib) due to 6 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

petrochenkov commented:
This PR implements two independent changes
- The `TyKind::Anon{Struct,Union}` -> `FieldTy` refactoring, which probably make sense as a cleanup
- The support for unnamed fields in built-in derives, which is a functional change, and probably doesn't make much sense until the unnamed fields are actually implemented (in HIR and further).

Could you keep only the first part in this PR?
The second one can be submitted separately and later.
@rustbot author

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6523069471/job/17713323243) [(plain)](https://github.com/rust-lang/rust/commit/05c68b453258d8933e72dc5afc24c1d4a5094aa2/checks/17713323243/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=frank-king
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_9bc459a6-4d27-460f-bf20-2b330ea60857
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=feature/unnamed-fields-lite
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_9bc459a6-4d27-460f-bf20-2b330ea60857
GITHUB_REF=refs/pull/115732/merge
GITHUB_REF_NAME=115732/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=0eca899e4e5069dbc0f658fa8a97b794244c8a47
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_9bc459a6-4d27-460f-bf20-2b330ea60857
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_9bc459a6-4d27-460f-bf20-2b330ea60857
GITHUB_TRIGGERING_ACTOR=frank-king
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115732/merge
GITHUB_WORKFLOW_SHA=0eca899e4e5069dbc0f658fa8a97b794244c8a47
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
    Checking addr2line v0.21.0
error[E0425]: cannot find function `eq_field_ty` in this scope
   --> src/tools/clippy/clippy_utils/src/ast_utils.rs:559:12
    |
229 | pub fn eq_field(l: &ExprField, r: &ExprField) -> bool {
    | ----------------------------------------------------- similarly named function `eq_field` defined here
...
559 |         && eq_field_ty(l.ty.expect_ty(), r.ty.expect_ty())
    |            ^^^^^^^^^^^ help: a function with a similar name exists: `eq_field`
    Checking anstyle-parse v0.2.1
   Compiling color-print-proc-macro v0.3.5
    Checking tracing-error v0.2.0
   Compiling declare_clippy_lint v0.1.75 (/checkout/src/tools/clippy/declare_clippy_lint)
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6523107930/job/17713409993) [(plain)](https://github.com/rust-lang/rust/commit/7d3ee85289d8666308a9bdf27a605c47d156dc89/checks/17713409993/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=frank-king
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_93cc1b55-e956-46ab-b782-e136c28f28aa
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=feature/unnamed-fields-lite
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_93cc1b55-e956-46ab-b782-e136c28f28aa
GITHUB_REF=refs/pull/115732/merge
GITHUB_REF_NAME=115732/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=56f5bbc556e1091195f91912466e1664ec1c8c1d
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_93cc1b55-e956-46ab-b782-e136c28f28aa
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_93cc1b55-e956-46ab-b782-e136c28f28aa
GITHUB_TRIGGERING_ACTOR=frank-king
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115732/merge
GITHUB_WORKFLOW_SHA=56f5bbc556e1091195f91912466e1664ec1c8c1d
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
    Checking toml_edit v0.19.11
    Checking semver v1.0.17
    Checking cargo_metadata v0.15.4
    Checking toml v0.7.5
error[E0599]: no method named `rewrite` found for enum `FieldTy` in the current scope
     |
     |
1892 |         .and_then(|ty_shape| field.ty.rewrite(context, ty_shape));
     |                                       ^^^^^^^ method not found in `FieldTy`
     |
note: the method `rewrite` exists on the type `P<rustc_ast::Ty>`
     |
     |
18   |     fn rewrite(&self, context: &RewriteContext<'_>, shape: Shape) -> Option<String>;
     = help: items from traits can only be used if the trait is implemented and in scope
     = help: items from traits can only be used if the trait is implemented and in scope
note: `Rewrite` defines an item `rewrite`, perhaps you need to implement it
     |
     |
16   | pub(crate) trait Rewrite {

error[E0282]: type annotations needed for `std::option::Option<U>`
    --> src/tools/rustfmt/src/items.rs:1890:9
     |
     |
1890 |     let orig_ty = shape
     |         ^^^^^^^
...
1894 |         if !ty.contains('\n') && !contains_comment(context.snippet(missing_span)) {
     |                -------- type must be known at this point
     |
help: consider giving `orig_ty` an explicit type, where the type for type parameter `T` is specified
     |
1890 |     let orig_ty: std::option::Option<U> = shape


error[E0614]: type `FieldTy` cannot be dereferenced
     |
     |
1901 |     let field_str = rewrite_assign_rhs(context, prefix, &*field.ty, &RhsAssignKind::Ty, shape)?;


error[E0615]: attempted to take value of method `span` on type `FieldTy`
    |
    |
141 |         span_with_attrs_lo_hi!(self, self.span.lo(), self.ty.span.hi())
    |
help: use parentheses to call the method
    |
    |
141 |         span_with_attrs_lo_hi!(self, self.span.lo(), self.ty.span().hi())


error[E0599]: no variant or associated item named `AnonStruct` found for enum `TyKind` in the current scope
    |
    |
822 |             ast::TyKind::AnonStruct(_) => Some(context.snippet(self.span).to_owned()),
    |                          ^^^^^^^^^^ variant or associated item not found in `TyKind`

error[E0599]: no variant or associated item named `AnonUnion` found for enum `TyKind` in the current scope
    |
    |
823 |             ast::TyKind::AnonUnion(_) => Some(context.snippet(self.span).to_owned()),
    |                          ^^^^^^^^^ variant or associated item not found in `TyKind`
Some errors have detailed explanations: E0282, E0599, E0614, E0615.
For more information about an error, try `rustc --explain E0282`.
error: could not compile `rustfmt-nightly` (lib) due to 6 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #116688) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
@frank-king 
Could you set the `S-waiting-on-review` label when this PR is rebased and ready?
It can be done with `@rustbot ready` and [other similar commands](https://rustc-dev-guide.rust-lang.org/contributing.html#waiting-for-reviews).

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6865078199/job/18668244768) [(plain)](https://github.com/rust-lang/rust/commit/353d71ff423e2e7b7c286a780f03e9a39aa9ccb1/checks/18668244768/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=frank-king
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_b9bdf9be-c2ae-4e87-86d2-ccab44d0e3e3
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=feature/unnamed-fields-lite
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_b9bdf9be-c2ae-4e87-86d2-ccab44d0e3e3
GITHUB_REF=refs/pull/115732/merge
GITHUB_REF_NAME=115732/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=50b1eccdba49315bcc8518c1b0099bc5417167ad
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_b9bdf9be-c2ae-4e87-86d2-ccab44d0e3e3
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_b9bdf9be-c2ae-4e87-86d2-ccab44d0e3e3
GITHUB_TRIGGERING_ACTOR=frank-king
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115732/merge
GITHUB_WORKFLOW_SHA=50b1eccdba49315bcc8518c1b0099bc5417167ad
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container 4de9bf3420b4
 ---> 879eb64ba64c
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 09befe65c7b0
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-dno5ab6p/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
  Downloading virtualenv-20.24.6-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 63.3 MB/s eta 0:00:00
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 127.3 MB/s eta 0:00:00
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.13.1 platformdirs-3.11.0 virtualenv-20.24.6
Removing intermediate container 95387b289258
 ---> 4e32e584cbfd
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 5ddb568ee94b
 ---> 5ddb568ee94b
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 6f545debf320
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container d25e6940db4f
 ---> cfe24b2336fb
Successfully built cfe24b2336fb
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:cfe24b2336fb55d62dea837248bd48d684a08b2120f2f125dd8de3e2e2bc31ac
Uploading finished image sha256:cfe24b2336fb55d62dea837248bd48d684a08b2120f2f125dd8de3e2e2bc31ac to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
cfe24b2336fb   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
5ddb568ee94b   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
4e32e584cbfd   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.8MB    
09befe65c7b0   9 seconds ago    /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
879eb64ba64c   10 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      5 weeks ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      5 weeks ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      5 weeks ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7fd55c25b890>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:cfe24b2336fb55d62dea837248bd48d684a08b2120f2f125dd8de3e2e2bc31ac
---
DirectMap4k:      171968 kB
DirectMap2M:     7168000 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/fa14810f21f47aa55ae7b3dd7c6d32c41e2072f8/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-fa14810f21f47aa55ae7b3dd7c6d32c41e2072f8-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 26.04s
##[endgroup]
fmt check
error[internal]: left behind trailing whitespace
    |
    |
206 |         if let Some(ident) = field.ident 
    |


error[internal]: left behind trailing whitespace
    |
    |
207 |             && ident.name == kw::Underscore 
    |


warning: rustfmt has failed to format. See previous 2 errors.

Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/bootstrap/src/utils/bin_helpers.rs" "/checkout/src/bootstrap/src/utils/dylib.rs" "/checkout/compiler/rustc_ast_passes/src/feature_gate.rs" "/checkout/compiler/rustc_ast_passes/src/ast_validation.rs" "/checkout/compiler/rustc_ast_passes/src/errors.rs" "/checkout/compiler/rustc_ast_passes/src/lib.rs" "/checkout/compiler/rustc_ast_passes/src/show_span.rs" "/checkout/src/bootstrap/src/utils/mod.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Tue Nov 14 14:37:56 UTC 2023
  network time: Tue, 14 Nov 2023 14:37:57 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #117915) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6919338014/job/18822587029) [(plain)](https://github.com/rust-lang/rust/commit/6da4302a27036782d24903dd78ca8808e5713973/checks/18822587029/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=frank-king
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_44c09409-e7ef-4a62-9040-ee670d2f1ac4
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=feature/unnamed-fields-lite
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_44c09409-e7ef-4a62-9040-ee670d2f1ac4
GITHUB_REF=refs/pull/115732/merge
GITHUB_REF_NAME=115732/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=bb2b9be8bb580829b7ac78829cefada7764e4c5e
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_44c09409-e7ef-4a62-9040-ee670d2f1ac4
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_44c09409-e7ef-4a62-9040-ee670d2f1ac4
GITHUB_TRIGGERING_ACTOR=frank-king
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115732/merge
GITHUB_WORKFLOW_SHA=bb2b9be8bb580829b7ac78829cefada7764e4c5e
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container e56012e8678e
 ---> 9790b0e958ab
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 84059acb400c
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-ymo8f3_h/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.24.6-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 58.7 MB/s eta 0:00:00
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.13.1 platformdirs-3.11.0 virtualenv-20.24.6
Removing intermediate container a086ba9fb279
 ---> 885b2f9946c9
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> c0785c458c72
 ---> c0785c458c72
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 4c2ce55b3cda
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 603d19c932ac
 ---> 6f1436356da7
Successfully built 6f1436356da7
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:6f1436356da78498003aca8f335943ea4065262b860869226098da3a1f80a641
Uploading finished image sha256:6f1436356da78498003aca8f335943ea4065262b860869226098da3a1f80a641 to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
6f1436356da7   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
c0785c458c72   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
885b2f9946c9   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.8MB    
84059acb400c   10 seconds ago   /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
9790b0e958ab   11 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      6 weeks ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      6 weeks ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      6 weeks ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7fc132dcb9d0>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:6f1436356da78498003aca8f335943ea4065262b860869226098da3a1f80a641
---
DirectMap4k:      174016 kB
DirectMap2M:     7165952 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/7d0e1bca0fd1cf26cb71a1271a782cd66ea1d9a9/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-7d0e1bca0fd1cf26cb71a1271a782cd66ea1d9a9-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 25.04s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_ast_lowering/src/item.rs at line 722:
             FieldTy::Ty(ty) => {
                 self.lower_ty(ty, &ImplTraitContext::Disallowed(ImplTraitPosition::FieldTy))
             }
-            FieldTy::AnonRecord(anon_record ) => {
+            FieldTy::AnonRecord(anon_record) => {
                 let struct_or_union = anon_record.kind.name();
                 let hir_id = self.lower_node_id(anon_record.id);
                 let span = anon_record.span;
##[error]Diff in /checkout/compiler/rustc_ast_lowering/src/item.rs at line 729:
                 // FIXME(unnamed_fields): IMPLEMENTATION IN PROGRESS
                 #[allow(rustc::untranslatable_diagnostic)]
                 #[allow(rustc::diagnostic_outside_of_impl)]
-                let kind = hir::TyKind::Err(self.tcx.sess.span_err(span, format!("anonymous {struct_or_union}s are unimplemented")));
+                let kind = hir::TyKind::Err(
+                    self.tcx
+                        .sess
+                        .span_err(span, format!("anonymous {struct_or_union}s are unimplemented")),
+                );
                 let ty = hir::Ty { hir_id, kind, span };
                 self.arena.alloc(ty)
             }
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_trait_selection/src/traits/select/confirmation.rs" "/checkout/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs" "/checkout/compiler/rustc_trait_selection/src/traits/select/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/project.rs" "/checkout/compiler/rustc_llvm/build.rs" "/checkout/compiler/rustc_resolve/src/imports.rs" "/checkout/compiler/rustc_ast_lowering/src/item.rs" "/checkout/compiler/rustc_trait_selection/src/traits/vtable.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Sun Nov 19 08:27:00 UTC 2023
  network time: Sun, 19 Nov 2023 08:27:01 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6919496235/job/18822939907) [(plain)](https://github.com/rust-lang/rust/commit/f711b6be7048192710b5af4229f12f1cacf0b19e/checks/18822939907/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=frank-king
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_da2de14c-8245-435a-8ba5-68079c67a6db
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=feature/unnamed-fields-lite
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_da2de14c-8245-435a-8ba5-68079c67a6db
GITHUB_REF=refs/pull/115732/merge
GITHUB_REF_NAME=115732/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=65142943243e789f6e709b0b2d72c05ceab3c942
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_da2de14c-8245-435a-8ba5-68079c67a6db
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_da2de14c-8245-435a-8ba5-68079c67a6db
GITHUB_TRIGGERING_ACTOR=frank-king
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115732/merge
GITHUB_WORKFLOW_SHA=65142943243e789f6e709b0b2d72c05ceab3c942
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container 787cd301a55b
 ---> f87df78c173e
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 6e0b3bc7dbba
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-dj12i5zv/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.24.6-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 19.0 MB/s eta 0:00:00
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.13.1 platformdirs-3.11.0 virtualenv-20.24.6
Removing intermediate container ac3a7cb33f4b
 ---> 0b70e7bbb449
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 8d2e24f031e8
 ---> 8d2e24f031e8
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> d7b92028ea46
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 7b4dfb139ead
 ---> dee65faddaba
Successfully built dee65faddaba
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:dee65faddaba91c47b9199e47354a95f9e5196ccd26da67afb7e8aae19279e67
Uploading finished image sha256:dee65faddaba91c47b9199e47354a95f9e5196ccd26da67afb7e8aae19279e67 to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED              CREATED BY                                      SIZE      COMMENT
dee65faddaba   1 second ago         /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
8d2e24f031e8   2 seconds ago        /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
0b70e7bbb449   4 seconds ago        |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.8MB    
6e0b3bc7dbba   11 seconds ago       /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
f87df78c173e   13 seconds ago       |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      6 weeks ago          /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      6 weeks ago          /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      6 weeks ago          /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f525d43b850>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:dee65faddaba91c47b9199e47354a95f9e5196ccd26da67afb7e8aae19279e67
---
processor : 0
vendor_id : AuthenticAMD
cpu family : 25
model  : 1
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
stepping : 1
microcode : 0xffffffff
cpu MHz  : 3152.725
cache size : 512 KB
---
DirectMap4k:      253888 kB
DirectMap2M:     8134656 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/7d0e1bca0fd1cf26cb71a1271a782cd66ea1d9a9/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-7d0e1bca0fd1cf26cb71a1271a782cd66ea1d9a9-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 25.00s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_parse/src/parser/item.rs at line 1776:
 
                 // Try to recover extra trailing angle brackets
                 let mut recovered = false;
-                if let FieldTy::Ty(ty) = &a_var.ty && let TyKind::Path(_, Path { segments, .. }) = &ty.kind {
+                if let FieldTy::Ty(ty) = &a_var.ty
+                    && let TyKind::Path(_, Path { segments, .. }) = &ty.kind
                     if let Some(last_segment) = segments.last() {
                         recovered = self.check_trailing_angle_brackets(
                             last_segment,
                             last_segment,
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_parse/src/lexer/tokentrees.rs" "/checkout/compiler/rustc_parse/src/lexer/unescape_error_reporting.rs" "/checkout/compiler/rustc_parse/src/validate_attr.rs" "/checkout/compiler/rustc_parse/src/parser/item.rs" "/checkout/compiler/rustc_driver/src/lib.rs" "/checkout/compiler/rustc_lint/src/noop_method_call.rs" "/checkout/compiler/rustc_parse_format/src/tests.rs" "/checkout/compiler/rustc_parse/src/parser/pat.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Sun Nov 19 09:05:58 UTC 2023
  network time: Sun, 19 Nov 2023 09:05:58 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6919588578/job/18823126644) [(plain)](https://github.com/rust-lang/rust/commit/6670134a3d4634d4dcfb5ec1ce75d2f8f5f77e99/checks/18823126644/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=frank-king
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_93151730-eb21-40ff-951b-f523f87c2328
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=feature/unnamed-fields-lite
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_93151730-eb21-40ff-951b-f523f87c2328
GITHUB_REF=refs/pull/115732/merge
GITHUB_REF_NAME=115732/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=c801f7d1adf0f98344273f9689e224eb2f335b27
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_93151730-eb21-40ff-951b-f523f87c2328
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_93151730-eb21-40ff-951b-f523f87c2328
GITHUB_TRIGGERING_ACTOR=frank-king
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115732/merge
GITHUB_WORKFLOW_SHA=c801f7d1adf0f98344273f9689e224eb2f335b27
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container 43d97406d65f
 ---> 9d4785673cf3
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 4bb5eb840a70
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-ngo20od7/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
  Downloading virtualenv-20.24.6-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 15.3 MB/s eta 0:00:00
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 34.3 MB/s eta 0:00:00
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.13.1 platformdirs-3.11.0 virtualenv-20.24.6
Removing intermediate container 047ef73e631a
 ---> d4179bcb8f12
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 606dcdb71225
 ---> 606dcdb71225
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> cd2caa22f85c
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 3c375917a4cc
 ---> 094c831ca6c6
Successfully built 094c831ca6c6
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:094c831ca6c6bbb5aafe0f5888c41d400d2d704ad81c182680e03ce8ff3285b9
Uploading finished image sha256:094c831ca6c6bbb5aafe0f5888c41d400d2d704ad81c182680e03ce8ff3285b9 to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED              CREATED BY                                      SIZE      COMMENT
094c831ca6c6   1 second ago         /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
606dcdb71225   2 seconds ago        /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
d4179bcb8f12   3 seconds ago        |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.8MB    
4bb5eb840a70   11 seconds ago       /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
9d4785673cf3   12 seconds ago       |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      6 weeks ago          /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      6 weeks ago          /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      6 weeks ago          /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7fd4715e75d0>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:094c831ca6c6bbb5aafe0f5888c41d400d2d704ad81c182680e03ce8ff3285b9
---
##[group]Display CPU and Memory information
processor : 0
vendor_id : AuthenticAMD
cpu family : 25
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
model name : AMD EPYC 7763 64-Core Processor
stepping : 1
microcode : 0xffffffff
cpu MHz  : 3243.094
---
DirectMap4k:      268224 kB
DirectMap2M:     6023168 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.05s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/7d0e1bca0fd1cf26cb71a1271a782cd66ea1d9a9/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-7d0e1bca0fd1cf26cb71a1271a782cd66ea1d9a9-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 25.10s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_ast_passes/src/ast_validation.rs at line 206:
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_apple_darwin.rs" "/checkout/compiler/rustc_target/src/tests.rs" "/checkout/compiler/rustc_target/src/lib.rs" "/checkout/compiler/rustc_target/src/json.rs" "/checkout/compiler/rustc_ast_passes/src/feature_gate.rs" "/checkout/compiler/rustc_target/src/abi/call/amdgpu.rs" "/checkout/compiler/rustc_ast_passes/src/ast_validation.rs" "/checkout/compiler/rustc_target/src/asm/csky.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
         if let Some(ident) = field.ident
             && ident.name == kw::Underscore
         {
-                let ty = self.check_unnamed_field_ty(&field.ty, ident.span);
-                self.visit_vis(&field.vis);
-                self.visit_ident(ident);
-                if let Some(ty) = ty {
-                    self.visit_ty_common(&ty);
-                    self.walk_ty(&ty);
-                }
-                walk_list!(self, visit_attribute, &field.attrs);
+            let ty = self.check_unnamed_field_ty(&field.ty, ident.span);
+            self.visit_vis(&field.vis);
+            self.visit_ident(ident);
+            if let Some(ty) = ty {
+                self.visit_ty_common(&ty);
+                self.walk_ty(&ty);
+            }
+            walk_list!(self, visit_attribute, &field.attrs);
         } else {
             self.visit_field_def(field);
Build completed unsuccessfully in 0:00:47
  local time: Sun Nov 19 09:22:13 UTC 2023
  network time: Sun, 19 Nov 2023 09:22:13 GMT
##[error]Process completed with exit code 1.
```

</details>

frank-king commented:
@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #118143) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

frank-king committed:
Parse unnamed fields and anonymous structs or unions

Anonymous structs or unions are parsed as a special type of field,
instead of a gerenal kind of `Ty`.

frank-king commented:
The problem is that, since `AnonStruct` and `AnonUnion` is not regarded as a `TyKind` at all, everywhere it should be reconsidered as a special kind of struct field where it appears.

The current approach (before this PR) requires more efforts on parsing anonymous structs or unions as a special `TyKind`, but after PR, we need more efforts on the special `FieldTy`, which is the side effect.

petrochenkov commented:
Yeah, I'm not convinced such refactoring is an improvement.
Everything becomes more complex, and even ast_validation.rs doesn't become simpler.

Work on RFC 2102 needs to rather focus on implementing the feature further in middle end and code generation.
Maybe after the feature is complete we can consider refactoring the representation to fit it better.

I'm going to close this in meantime.
